mpir: add front-end support for parallel debugging

dongahn opened the pull request:
This PR is work in progress. At least three things are required before it can be merged.
- `flux-info` module provides a single job query, an equivalent of ba95bf6 and 
6068c24. Tagging @chu11 and @grondo for this as this is their planned task.
- I need to find a way to disable compilers to optimize (e.g., inlining) `MPIR_Breakpoint` (fec88d6) so that this function can remain empty as it should. (Currently, I added two printf statements lest totalview cannot reliably get its MPIR_Breakpoint event). Tagging @jdelsign in case he has some insight. 
- Test this on the LC environment as I did my manual totalview testing under my docker development environment. 

Having said that, this PR adds MPIR front-end support for parallel debuggers like totalview.

- Add the MPIR debug interface within `flux-job`’s `attach` command.
- Add an automatic debugger-attach detection scheme into `flux-mini.py` so that users don't have to specify `-o stop-tasks-in-exec` when they launch a job using `flux-mini run` under debugger control. To achieve this, a very thin debugger-attach detection library (`libdebugged`) has been added and the python binding has been extended to include that.
- Add tests, in particular a sharness test that drives this logic in an emulation environment. For this purpose, `--debug-emulate` option has been added to both the `flux-mini run` command as well as `flux-job attach` command. 

If anyone wants to do some manual testing with totalview:

For launch mode:
% totalview-install/toolworks/totalview.2019.3.14/bin/totalview -oldui -verbosity error --args flux mini run -N 2 -n 2 ./sleep_print 60 5

For attach mode:
totalview-install/toolworks/totalview.2019.3.14/bin/totalview -oldui -verbosity error --args flux job attach <JOBID>

I also tested both modes with subset attach whereby totalview only attaches to a subset of processes chosen. And this worked well, hence we define the `MPIR_partial_attach_ok` variable in `flux-job`.
 
Resolve #2597 and #2538.

jdelsign commented:
>     * I need to find a way to disable compilers to optimize (e.g., inlining) `MPIR_Breakpoint` ([fec88d6](https://github.com/flux-framework/flux-core/commit/fec88d68763048f88d68c676e91749a0284d7de1)) so that this function can remain empty as it should. (Currently, I added two printf statements lest totalview cannot reliably get its MPIR_Breakpoint event). Tagging @jdelsign in case he has some insight.

As I'm sure you know, defeating compiler optimization can be tricky, which is one of the reasons MPIR is problematic. In any case, the following comes to mind:
* Isolate the MPIR stuff in a separate module and compile it w/out optimization.
* Take the address of the function and stick it in a variable.

grondo commented:
@dongahn: just started looking at this, but a couple things at first:

 - `src/common/libdebugged/Makefile.am` seems to be missing
 - Python code needs to be formatted with `black` (some issues with whitespace in `flux-mini.py` will need to be fixed first.

SteVwonder commented:
Very nice @dongahn!! The changes to the bindings generator LGTM!

grondo commented on src/bindings/python/_flux/Makefile.am:
```diff
@@ -31,7 +33,7 @@ _core_build.py: $(MAKE_BINDING)
 		--additional_headers src/bindings/python/_flux/callbacks.h \
 		--ignore_header 'macros' \
-		src/include/flux/core.h
+		src/include/flux/core.h src/include/flux/debugged.h
```
Isn't `debugged.h` in `src/common/libdebugged/debugged.h`, not `src/include/flux/debugged.h`?

dongahn replied:
`src/include/flux/debugged.h` is there as well, in which I included `src/common/libdebugged/debugged.h`. Maybe I didn't check that in? I will check and add that file if not committed in.

grondo replied:
Oh, is libdebugged/debugged.h part of the public API?

dongahn replied:
> Oh, is libdebugged/debugged.h part of the public API?

Depends what you mean by this. But it needs that to make python binding code a bit simple by allowing. 

```python
   for header in absolute_heads:
        process_header(header)
        includes += "#include <{}>\n".format(header)
```

Also I was to try to link it with `flux-job` to see if that will disallow compilers to use optimization on MPIR_Breakpoint function.

Does it make it public enough?

grondo replied:
Hm, I'm probably not familiar enough with the python cffi stuff to understand how this works, but for an installed flux-core, how will the python `flux.debugged` module get the symbols from `libdebugged`?

Instead of being a cffi module could `flux.debugged` be a .so module compiled against libdebugged.la convenience library so that libdebugged.la and its symbols don' t need to be added to libflux-core.so?

In either case, I don't think is should be necessary to install `/usr/include/flux/debugged.h` if that isn't meant to be part of the libflux api (just part of the python api), so perhaps `src/include/debugged.h` wrapper isn't required?

dongahn replied:
The symbols will be first included into `_core` and `flux.debugged` will get them from there. So the MPIR symbols will not be included into `libflux-core.so`. 

I actually agree this header file don't need to be installed in the public location. 

Maybe I should change `make_python.py` to use double quotes (`"`) instead of angle brackets (`'<'` etc) for `#include` used with `ffi.cdef()`. If this works, we should be good. I will test this but @SteVwonder or @trws will ultimately have to review it.

grondo replied:
Ah, ok that sounds perfect then!

grondo commented on src/cmd/Makefile.am:
```diff
@@ -20,6 +20,7 @@ fluxcmd_ldadd = \
 	$(top_builddir)/src/common/libflux-internal.la \
 	$(top_builddir)/src/common/libflux-core.la \
 	$(top_builddir)/src/common/libflux-optparse.la \
+        $(top_builddir)/src/shell/libmpir.la \
```
`libmpir.la` should be added to only the `flux-job` command, not all commands. (See `flux_start_LDADD` for an example of how to do this)

dongahn replied:
Yes you are right. Sorry about that.

grondo commented on src/cmd/flux-job.c:
```diff
@@ -134,6 +158,9 @@ static struct optparse_option attach_opts[] =  {
       .usage = "Suppress warnings written to stderr from flux-job",
     },
+    { .name = "debug-emulate", .key = 'd', .has_arg = 0,
+      .usage = "Set MPIR_being_debugged for testing",
```
We probably don't want to take up a short option `-d` for a test-only option. 

Also, consider `.flags = OPTPARSE_OPT_HIDDEN` so the option doesn't appear in `flux job attach --help` otuput.

dongahn replied:
Ah... I was wondering why some options have short and long options while others only long. This explains it. I will change it (along with `OPTPARSE_OPT_HIDDEN`).

grondo commented on src/cmd/flux-job.c:
```diff
@@ -1158,6 +1185,124 @@ void attach_stdin_cb (flux_reactor_t *r, flux_watcher_t *w,
+                                "t_submit",
+                                &t_submit,
+                                "state",
+                                &state) < 0)
```
Style: for pack/unpack, we seem to have settled on a style that collects key, value on the same line, e.g.:
```C
    if (flux_rpc_get_unpack (f, "{s:I s:i s:i s:f s:i}",
￼                                "id", &id,
￼                                "userid", &userid,
￼                                "priority", &priority,
￼                                "t_submit", &t_submit,
￼                                "state", &state) < 0)
```

However, that's not official so probably final style is up to you.

Edit: Same comment for `flux_rpc_pack()` above.

dongahn replied:
Sure. I can make that change. This may change in the future when this will get this from `job-info`. But it would hurt to do this now as a reminder.

grondo replied:
Yeah, that was just meant as an FYI :wink:

dongahn replied:
Yeah but this FYI is good!

I think it makes sense to change it (as well as fetching only state) now so that when `job-info` comes hopefully we can make one change and we can land this.  :-)

grondo commented on src/cmd/flux-job.c:
```diff
@@ -1158,6 +1185,124 @@ void attach_stdin_cb (flux_reactor_t *r, flux_watcher_t *w,
+
+    flux_future_destroy (f);
+
+    if (state > FLUX_JOB_RUN)
```
Minor: Since only `state` is used in this function, you could change the `unpack` above to only unpack the `state`, i.e. you are not required to unpack all contents of a message if you don't need them. This also makes the unpack less sensitive to changes in the message payload.

grondo replied:
Also, as @garlick pointed out in #2635, this comparison might be a bit dangerous since a job state `> FLUX_JOB_RUN` could be added in the future which actually precedes the RUN state.

A better test here would to see ensure state is `FLUX_JOB_PENDING | FLUX_JOB_RUN`

dongahn replied:
Good point. I can change this. 

I assumed the enum numeric values monotonically increases as a function of job lifecycle. (except for cyclic state transitions). Don't we have enough number spacing in the existing enum fields though? So that future states can also be added in lifecycle order?

grondo replied:
A good question. The states seem to be treated as flags so I'm not sure the intent there.
```C
typedef enum {
    FLUX_JOB_NEW                    = 1,
    FLUX_JOB_DEPEND                 = 2,
    FLUX_JOB_SCHED                  = 4,
    FLUX_JOB_RUN                    = 8,
    FLUX_JOB_CLEANUP                = 16,
    FLUX_JOB_INACTIVE               = 32,   // captive end state
} flux_job_state_t;
```
Maybe @garlick can comment on the design of `flux_job_state_t`

However, I don't think there is any guarantee of order implied, and in the future the state diagram may not be completely linear anyway. E.g. a STOP or similar state may be both before and after RUN.

dongahn replied:
Yeah, like I said I'm fine with making this change.  

Going forward, I think we will have to have cycles with grow and shrink... It seems it is a good practice to have some "loose ordering" for readability though.

garlick replied:
They are consecutive powers of two (to make them maskable), so no actually.  Also just FYI but [RFC 21](https://github.com/flux-framework/rfc/blob/master/spec_21.adoc) says:
> The job state machine SHALL NOT contain cycles, to avoid ambiguity in synchronization.

garlick replied:
ooops sorry - I didn't see your other comments when I posted mine.   I thought we had talked through the issue of cycles and grow/shrink back when RFC 21 was landed.  I'll go see if there's a discussion over there somewhere and open an issue if not.

grondo replied:
Ah, thanks @garlick! I had forgotten about that rule in the RFC. And sorry about the dumb question about state masks (if I had looked at the virtual job states again I would have understood)

grondo commented on src/cmd/flux-mini.py:
```diff
@@ -176,6 +177,12 @@ def create_parser(self):
         )
+        parser.add_argument(
+            "-d",
+            "--debug-emulate",
```
Consider making this a long-only option, and hide it from default output with `help=argparse.SUPPRESS`

dongahn replied:
Sure.

dongahn commented:
@jdelsign: 

Thanks John.

> Take the address of the function and stick it in a variable.

I tried this but didn't work. 
 
> Isolate the MPIR stuff in a separate module and compile it w/out optimization.

Yeah this was my next idea too. I will try this.

dongahn commented:
> src/common/libdebugged/Makefile.am seems to be missing

Yeah, thank you for catching. It is in my local repo, but didn't make in. I will add them.


> Python code needs to be formatted with black (some issues with whitespace in flux-mini.py will need to be fixed first.

I just noticed that from Travis as well. Good protocol to remember. Will do and commit a change.

dongahn commented:
I think I incorporated all of the suggested changes. However, I am getting testing errors (maybe a hang leading to Travis CI timeout). One example is the following although there are other variants. 

Anyone seen this? 

```
t2204-job-info.t:  PASS: N=21  PASS=21  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
4051ERROR: t2204-job-info.t - missing test plan
4052ERROR: t2204-job-info.t - exited with status 137 (terminated by signal 9?)
4053      t2300-sched-simple.t:  PASS: N=23  PASS=23  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
4054t2301-schedutil-outstanding-requests.t:  PASS: N=7   PASS=7   FAIL=0 SKIP=0 XPASS=0 XFAIL=0
4055     t2400-job-exec-test.t:  PASS: N=12  PASS=12  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
4056    t2401-job-exec-hello.t:  PASS: N=10  PASS=10  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
4057    t2402-job-exec-dummy.t:  PASS: N=12  PASS=12  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
4058        t2500-job-attach.t:  PASS: N=11  PASS=11  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
4059   t2600-job-shell-rcalc.t:  PASS: N=13  PASS=13  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
4060t2601-job-shell-standalone.t:  PASS: N=32  PASS=32  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
```

dongahn commented:
When I ran `make check` on my Mac OSX docker, test 22 seems to hang...

```console
ahn1@035029f1d7bf:/usr/src/t$ ./t2204-job-info.t
ok 1 - job-info: generate jobspec for simple test job
ok 2 - load job-exec,sched-simple modules
ok 3 - submit jobs for job list testing
ok 4 - flux job list running jobs in started order
ok 5 - flux job list running jobs with correct state
ok 6 - flux job list inactive jobs in completed order
ok 7 - flux job list inactive jobs with correct state
ok 8 - flux job list pending jobs in priority order
ok 9 - flux job list pending jobs with correct priority
ok 10 - flux job list pending jobs with correct state
ok 11 - flux job list jobs with correct userid
ok 12 - flux job list defaults to listing pending & running jobs
ok 13 - flux job list --userid=userid works
ok 14 - flux job list --userid=all works
ok 15 - flux job list --count works
ok 16 - flux job list all jobs works
ok 17 - job stats lists jobs in correct state (mix)
ok 18 - cleanup job listing jobs
ok 19 - reload the job-info module
ok 20 - job-info: list successfully reconstructed
ok 21 - job stats lists jobs in correct state (all inactive)
```

@grondo or @chu11: Is it possible there is a timing issue between `wait-event` and `flux job list -s inactive` or similar? 

```shell
# simply test that value in timestamp increases through job states
test_expect_success HAVE_JQ 'flux job list job state timing outputs valid (job inactive)' '
        jobid=$(flux mini submit hostname) &&
        flux job wait-event $jobid clean >/dev/null &&
        obj=$(flux job list -s inactive | grep $jobid) &&
        echo $obj | jq -e ".t_depend < .t_sched" &&
        echo $obj | jq ".t_sched < .t_run" &&
        echo $obj | jq ".t_run < .t_cleanup" &&
        echo $obj | jq ".t_cleanup < .t_inactive"
'
```

chu11 commented:
@dongahn (Sorry, didn't see you had tagged me in your recent comment) this could be potentially racy, as discussed in #2592.

grondo commented:
@dongahn, now that #2720 is merged I experimentally rebased this on current master, removed the new `job-manager.info` service and switched `flux job attach` to use `job-info.list-id` instead. Check it out here:

https://github.com/grondo/flux-core/tree/mpir_frontend

dongahn commented:
> @dongahn, now that #2720 is merged I experimentally rebased this on current master, removed the new job-manager.info service and switched flux job attach to use job-info.list-id instead. Check it out here:

> https://github.com/grondo/flux-core/tree/mpir_frontend

@grondo: LGTM. There is one place where `%ju` is used without a type casting (`uintmax_t`) to `ctx->id`. Other than this, this looks good to me. Let me incorporate these changes into this PR and finish this up. This PR has been open long enough. I think it makes sense to squash those intermediate commits as well and have the one final review pass. Let me know.

grondo commented:
I have a couple pending fixes. Running down one last problem (`make distcheck` fails)

dongahn commented:
> I have a couple pending fixes. Running down one last problem (make distcheck fails)

Good. If you push that in, I will cherry-pick them up along with other two fixes you had.

grondo commented:
Just force-pushed over my branch. The `make distcheck` issue I had was the python DESTDIR thing, so hopefully just occurring on my local copy.

dongahn commented:
@grondo: sorry it has taken so long to get to this. Your changes look good to me. One very minor comment:

https://github.com/flux-framework/flux-core/commit/8d0ff413a759daf3242d8e8971339345ac679904

would it be better to use `PATH_MAX` instead if this buffer is used to store a path? It is typically defined to be 4K, but it may give a better platform guarantee?

grondo commented:
It is just meant to be a "buffer of reasonable size". I shy away from
PATH_MAX lately because:

http://insanecoding.blogspot.com/2007/11/pathmax-simply-isnt.html

On Wed, Feb 26, 2020, 4:58 PM Dong H. Ahn <notifications@github.com> wrote:

> @grondo <https://github.com/grondo>: sorry it has taken so long to get to
> this. Your changes look good to me. One very minor comment:
>
> 8d0ff41
> <https://github.com/flux-framework/flux-core/commit/8d0ff413a759daf3242d8e8971339345ac679904>
>
> would it be better to use PATH_MAX instead if this buffer is used to
> store a path? It is typically defined to be 4K, but it may give a better
> platform guarantee?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2654?email_source=notifications&email_token=AAFVEUXOVCKPAJH7VHHIGE3RE4FZ5A5CNFSM4KIOB5RKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENCPLXI#issuecomment-591721949>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFVEURQUOIKN2DALH4HSWTRE4FZ5ANCNFSM4KIOB5RA>
> .
>

dongahn commented:
I rebased this to upstream/master and found an issue. This wasn't the case with the old flux-core code base. 

```
ahn1@d19bd615353c:/usr/src$ flux mini run -N 2 -n 2 sleep 0
<Hit Return>
flux-job: stdin not accepted by job
```

It seems 42f7bc19576f59780fb34ea9f767f06486f8dcd8 was the commit where this occurred. The job seemed to have run fine (confirmed with `flux kvs eventlog get`), but `flux mini run -N 2 -n 2 sleep 0` was never returning to the prompt. So when I  hit "Return" at which point the error message was printed. 

I will look into this further.

grondo commented:
I rebased my copy of your branch on current master and wasn't able to reproduce the issue. Make sure you pulled all the fixes from my branch into yours, especially this one:

```diff
commit 11a9840c225a082a28e87add0284f9bba70947e4
Author: Mark A. Grondona <mark.grondona@gmail.com>
Date:   Thu Feb 20 14:44:16 2020 +0000

    flux-mini: fix check for args.debug_emulate
    
    Problem: When checking for args.debug_emulate the check does
    
     if args.debug_emulate is not None:
    
    However, False is not None, so this argument is always enabled.
    
    Fix and just use `if args.debug_emulate`
---
 src/cmd/flux-mini.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/cmd/flux-mini.py b/src/cmd/flux-mini.py
index a5d0a6f..5d4fdfd 100755
--- a/src/cmd/flux-mini.py
+++ b/src/cmd/flux-mini.py
@@ -193,7 +193,7 @@ class SubmitCmd:
                     pass
                 jobspec.setattr_shell_option(key, val)
 
-        if args.debug_emulate is not None:
+        if args.debug_emulate:
             debugged.set_mpir_being_debugged(1)
 
         if debugged.get_mpir_being_debugged() == 1:

```

dongahn commented:
I did. 

I think there is one merge error in this commit per a refactored code made since I did this work. I will keep you posted.

dongahn commented:
> I think there is one merge error in this commit per a refactored code made since I did this work. I will keep you posted.

Yap. This was the case. I didn't remove the old output future code per a recent refactor. 

```c
        attach_output_start (ctx);
```

Sorry for the noise!

dongahn commented:
@grondo: Ok, I rebased it to the upstream/master and cherry-picked your changes and also squashed all the interim commits. The squashed commits included some of your commits as well. But it turned out git's squash didn't merge authorships... sorry about that.

 I ran some manual smoke tests with totalview on the docker environment on my Mac osx laptop and this worked fine as before.

There are a few enhancement I will need from Perforce totalview to streamline our parallel debugger support. As you can see from this PR, I have communicated this to Perforce. And I will create some tickets to track the requests.

I will see if Travis CI is happy with this PR tomorrow.

Once it turns green, it would be good if you can take a brief look at it before landing it.

grondo commented:
Well, the ASan builder hung, which has been happening lately. No way to tell which test hung, so I just restarted it.

grondo commented:
> The squashed commits included some of your commits as well. But it turned out git's squash didn't merge authorships... sorry about that.

No problem, I meant to say it was fine to squash any of my fixups into yours.

But, for future reference, when you are squashing commits by multiple authors, GitHub supports `Co-authored-by:` trailer as a standard way to support attribution.

https://github.blog/2018-01-29-commit-together-with-co-authors/

grondo commented:
ASan builder hung again, so I've restarted again. If another hang occurs I'll have to do some iterative debugging :cry:

dongahn commented:
> ASan builder hung again, so I've restarted again. If another hang occurs I'll have to do some iterative debugging

It could be something added to this PR. Is there a way to run this on my docker?

grondo commented:
> Is there a way to run this on my docker?

The config is documented in the `.travis.yml` for the `address-sanitizer` builder:
```yaml
    - name: "Ubuntu: gcc-8 address-sanitizer --with-flux-security"
      stage: test
      compiler: gcc-8
      env:
       - CC=gcc-8
       - CXX=g++-8
       - ARGS="--with-flux-security --enable-sanitizer=address"
       - PRELOAD=/usr/lib/gcc/x86_64-linux-gnu/8/libasan.so
       - ASAN_OPTIONS=detect_leaks=0 # TODO: fix/suppress tests so we can turn t
       - PYTHON_VERSION=3.6
```

So try:

```
$ CC=gcc-8 CXX=g++-8 PYTHON_VERSION=3.6 ./configure --enable-sanitizer=address
...
$ make -j 4
$ ASAN_OPTIONS=detect_leaks=0 LD_PRELOAD=/usr/lib/gcc/x86_64-linux-gnu/8/libasan.so make -j 4 check
```

grondo commented:
looks like asan builder completed this time though

jdelsign commented:
> There are a few enhancement I will need from Perforce totalview to streamline our parallel debugger support. As you can see from this PR, I have communicated this to Perforce. And I will create some tickets to track the requests.

@dongahn I can't seem to find these TotalView enhancement requests. Can you provide direct pointers to them? Thanks!

grondo commented:
@dongahn, is this ready to be merged? Thanks!

dongahn commented:
I think so. I don't plan to add more. Will force a push.

grondo commented:
>  Will force a push.

You don't have to worry about it since mergify will take care of the rebase (unless you have more work to add)

dongahn commented:
> You don't have to worry about it since mergify will take care of the rebase (unless you have more work to add)

Ah! Handy. Thanks.

grondo commented:
Thanks, I've added merge-when-passing label

grondo commented:
Hm, one builder failed in the new debug-emulate tests. Looks like the test hung somehow. It must be in the `flux job attach --debug-emulate ${jobid}` step since the `wait-event` test has a timeout.

We should probably try to reproduce this one.

```
debugger: SIGCONT can unlock a job in stop-tasks-in-exec
expecting success: 
        jobid=$(flux jobspec srun -n 2 ${stall} | flux job submit) &&
        flux job wait-event -vt 2.5 ${jobid} start &&
        flux job attach --debug-emulate ${jobid} &&
        flux job wait-event -vt 2.5 ${jobid} finish
1583172885.756965 submit userid=2000 priority=16 flags=0
1583172885.770071 depend
1583172885.773551 alloc note="rank0/core[0-1]"
1583172885.790271 start
0.065s: flux-shell[0]: ERROR: affinity: affinity: core1 not in topology
0.066s: flux-shell[0]:  WARN: affinity: unable to get cpuset for cores 0-1. Disabling affinity
298.123s: job.exception type=cancel severity=0 
flux-job: task(s) Killed
Caught SIGCONT
```

dongahn commented:
> 0.065s: flux-shell[0]: ERROR: affinity: affinity: core1 not in topology
0.066s: flux-shell[0]:  WARN: affinity: unable to get cpuset for cores 0-1. Disabling affinity

I have seen this when I used `hwloc` for testing... I have a meeting at 11AM so this will have to be this afternoon.

grondo commented:
You can ignore the affinity warnings, this is because the tests are using a fake resource.hwloc.by_rank configuration during testing, and will be present in all jobs.

The real problem is the hang.

dongahn commented:
> The real problem is the hang.

Let me try to reproduce this in my docker. 

BTW, the build badge in our README.md has "error" on it. Does anyone know why?
 
<img width="621" alt="Screen Shot 2020-03-04 at 5 37 40 PM" src="https://user-images.githubusercontent.com/862123/75939369-db647a80-5e3e-11ea-945e-c6aaa6c3bdd5.png">

grondo commented:
Latest build of flux-core master on travis-ci failed (asan builder hung, another one failed due to github error(?)) I've restarted the failed builders.

garlick commented:
So status of this PR is it's waiting for a fix to the "core1 not in topology" failure in https://github.com/flux-framework/flux-core/pull/2654#issuecomment-593556004?

dongahn commented:
The bigger problem is a hang, @garlick. I need to reproduce it to debug but I have kept being distracted...

grondo commented:
> So status of this PR is it's waiting for a fix to the "core1 not in topology" failure?

That isn't a failure, just a warning... It could be suppressed with a shell initrc config. Maybe I'll take a look at it

dongahn commented:
> Hm, one builder failed in the new debug-emulate tests. Looks like the test hung somehow. It must be in the flux job attach --debug-emulate ${jobid} step since the wait-event test has a timeout.

> We should probably try to reproduce this one.

@grondo: I don't seem to be able to reproduce this on my Mac/OSx Docker...

dongahn commented:
> debugger: SIGCONT can unlock a job in stop-tasks-in-exec
expecting success: 
        jobid=$(flux jobspec srun -n 2 ${stall} | flux job submit) &&
        flux job wait-event -vt 2.5 ${jobid} start &&
        flux job attach --debug-emulate ${jobid} &&
        flux job wait-event -vt 2.5 ${jobid} finish

Maybe there is a race condition in the test itself. If it takes some time for the target to install the `SIGCONT` signal handler, the `SIGCONT` signal sent by `flux job attach --debug-emulate ${jobid} ` is lost. Let me see if I can add some synchronization.

dongahn commented:
How about:

- Reduce the number of tasks for the target to 1.
- Let it write to the KVS right *after* the signal handler is installed 
- Then the sharness code waits for this KVS key before executing `flux job attach `...

dongahn commented:
I fixed the synchronization last night but now this test is failing with abort. (and ASAN hangs).

```
expecting success: 
6852        jobid=$(flux jobspec srun -n 1 ${stall} done 10 | flux job submit) &&
6853        flux job wait-event -vt 2.5 ${jobid} start &&
6854        ${waitfile} -v -t 2.5 done &&
6855        flux job attach --debug-emulate ${jobid} &&
6856        flux job wait-event -vt 2.5 ${jobid} finish
6857
68581583813900.177787 submit userid=2000 priority=16 flags=0
68591583813900.191253 depend
68601583813900.195665 alloc note="rank0/core0"
68611583813900.208126 start
6862waitfile: done: 0.000s: File appeared with size 0
6863waitfile: done: 0.000s: Got match (1/1)
6864waitfile: done: 0.000s: Got match. Calling self:on_match()
6865waitfile: done: 0.000s: Exiting
6866Aborted (core dumped)
6867not ok 5 - debugger: attaching to a running job works
```

I will have to run some Travis CI jobs. Bear with me while I will push a bunch of debug commits.

dongahn commented:
For some reason, `0.11s$ src/test/backtrace-all.sh` doesn't generate a backtrace for this abort, though.

grondo commented:
Yeah, they keep breaking core dumps in the travis-ci containers. Since this is the ASan builder, I wonder if there is some asan output, let me check.

dongahn commented:
I am not getting

```
1583875136.879217 submit userid=2000 priority=16 flags=0
68361583875136.894229 depend
68371583875136.901376 alloc note="rank[0-1]/core1"
68381583875136.934016 start
6839flux-job: shell-15267266560.proctable: Function not implemented
6840not ok 2 - debugger: submitting under debugger via flux-mini works
```

Probably from #2813...

garlick commented:
Yes that's the wrong service name.  It should come from the
guest.exec.eventlog like the rank.

On Wed, Mar 11, 2020, 12:27 PM Dong H. Ahn <notifications@github.com> wrote:

> I am now getting
>
> 1583875136.879217 submit userid=2000 priority=16 flags=0
> 68361583875136.894229 depend
> 68371583875136.901376 alloc note="rank[0-1]/core1"
> 68381583875136.934016 start
> 6839flux-job: shell-15267266560.proctable: Function not implemented
> 6840not ok 2 - debugger: submitting under debugger via flux-mini works
>
> Probably from #2813
> <https://github.com/flux-framework/flux-core/pull/2813>...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2654#issuecomment-597824531>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABJPWZBJHJF4UHQJEPWGRDRG7Q3DANCNFSM4KIOB5RA>
> .
>

grondo commented:
@dongahn, I have a little time to work on this. If you'd like I can fix up the conflicts and push to your branch. LMK.

dongahn commented:
@grondo: If you can do this, I will appreciate it. No time util the end of this week due to all-day meetings.

grondo commented:
Actually need some help from @trws or @SteVwonder here. When I update `make_clean_header.py` call to include `libdebugged.h`, and include `libdebugged.la` in `common_libs`, so that @dongahn's convenience library is ostensibly included in the bindings, it doesn't work:

```
make[1]: Entering directory '/home/grondo/git/flux-core.git/src/bindings/python/_flux'
  CC       _core.lo
_core.c: In function ‘_cffi_d_MPIR_Breakpoint’:
_core.c:2626:3: error: implicit declaration of function ‘MPIR_Breakpoint’; did you mean ‘_cffi_d_MPIR_Breakpoint’? [-Werror=implicit-function-declaration]
   MPIR_Breakpoint();
   ^~~~~~~~~~~~~~~
```

Only change so far is this:

What else is required to add new cffi bindings to the python stuff?

```diff
diff --git a/src/bindings/python/_flux/Makefile.am b/src/bindings/python/_flux/Makefile.am
index 33d447174..2ac8d0713 100644
--- a/src/bindings/python/_flux/Makefile.am
+++ b/src/bindings/python/_flux/Makefile.am
@@ -13,6 +13,7 @@ AM_LDFLAGS = \
 
 common_libs = $(top_builddir)/src/common/libflux-core.la \
              $(top_builddir)/src/common/libflux-internal.la \
+             $(top_builddir)/src/common/libdebugged/libdebugged.la \
              $(ZMQ_LIBS) $(LIBUUID_LIBS) \
              $(PYTHON_LDFLAGS)
 
@@ -24,7 +25,9 @@ _core_clean.h: Makefile
        $(PYTHON) $(srcdir)/make_clean_header.py \
          --path $(top_srcdir) \
          --search $(top_builddir)/src/common/libflux \
-         --additional_headers src/bindings/python/_flux/callbacks.h \
+         --additional_headers \
+             src/bindings/python/_flux/callbacks.h \
+             src/common/libdebugged/debugged.h \
          --output _core_clean.h \
          src/include/flux/core.h
 _core_preproc.h: _core_clean.h

```

grondo commented:
Hm, this works, but not sure if it is the correct appraoch. Why does the header file have to be specified twice?

```diff
diff --git a/src/bindings/python/_flux/_core_build.py b/src/bindings/python/_flux/_core_build.py
index fe5dc89bc..43285b109 100644
--- a/src/bindings/python/_flux/_core_build.py
+++ b/src/bindings/python/_flux/_core_build.py
@@ -7,6 +7,7 @@ ffi.set_source(
     "_flux._core",
     """
 #include <src/include/flux/core.h>
+#include <src/common/libdebugged/debugged.h>
 
 
 void * unpack_long(ptrdiff_t num){
```

trws commented:
The additional_headers flag tells the script which headers to read, but 
does not add an include of those headers in the set_source since it 
doesn’t modify the build file anymore, so that should be the right 
solution.

grondo commented:
Ok, I fixed the conflicts, squashed the fixup commits and force pushed to your branch @dongahn 

Tests were passing on my system. However, I'm confused because pylint has now decided to flag some errors in `flux/job.py`, though this PR doesn't touch that file (or shouldn't have)

grondo commented:
Ok, as discussed offline, there was an unused import error in `flux-mini.py`. The other errors existed previous to this PR run, but are currently ignored (to be fixed)

grondo commented:
Ok, looks like I've resolved the pylint error. I guess if this passes it will be merged because MWP is still set from before. Someone speak up if this shouldn't be merged I guess.. :see_no_evil:

dongahn commented:
@grondo: were you able to resolve: https://github.com/flux-framework/flux-core/pull/2654#issuecomment-597193987? Seems like it from the Travis CI. If so, I'm okay this to go in. Thank you so much for helping; the fact that I hadn't had a closure for this bothered me a lot while I'm fighting other assignments.

grondo commented:
Well this got merged into master since merge-when-passing was still set from before. I guess our mergify rule matching `[WIP]` doesn't seem to work.. hrm. Oh wel..
